libkmod/libkmod.c

```diff
@@ -810,22 +810,37 @@ KMOD_EXPORT void kmod_unload_resources(struct kmod_ctx *ctx)
 
 KMOD_EXPORT int kmod_dump_index(struct kmod_ctx *ctx, enum kmod_index type, int fd)
 {
+	int err = 0, fd2;
+	FILE *fp;
+
 	if (ctx == NULL)
 		return -ENOSYS;
 
+	fd2 = dup(fd);
+	if (fd2 == -1)
+		return -errno;
+	fp = fdopen(fd2, "ae");
+	if (fp == NULL) {
+		err = -errno;
+		close(fd2);
+		return err;
```

Comment:
Let's move all the whole dup/fdopen hunk after _all_ the input validation - aka after the `type` check below.

Reply:
Would adjust it if we decide if it's even worth it:

- FILE loses EINTR handling (modprobe -c shouldn't really care, since it first uses FILE and then passes a file descriptor to library, but let's not break API and whatever other users might rely on)
- wbuf speeds up processing, but introduces custom code, not improving the error handling

So ... let's see. I wouldn't mind keeping everything as it is right now and, if we ever adjust the API in a release, reconsider turning it FILE-based.

Reply:
I second that. What we could have is one additional API that receives a FILE* as argument.... we are not really going to change the API and break users because of this.

However if we add the new API, then we don't drop the old code since converting it to use FILE* would mean to lose EINTR handling. So.... I'm leaning towards the wbuf approach.